Configure ide (#11615)
mne-tools/mne-python: doc/install/configure_ide.rst

.. _configure-ide:

IDE Configuration Guide
=======================

.. include:: ../links.inc
.. highlight:: console

> Thanks for taking the time to contribute! This guide steps through various IDE settings that are helpful when
> contributing to MNE-Python.

Reviewer: ```suggestion
This guide steps through various IDE settings that are helpful when contributing to MNE-Python.
```
Given that this will land somewhere within the contributor's guide, I don't think we will need to add the thank you again.

Some of the basic settings include enabling linting, setting line margins, introspection, and docstring type:

> * Linting - is the automated checking of your source code for programmatic and stylistic errors. Most of the time basic linting is already enabled but let's double check.

Reviewer: ```suggestion
* Linting is the automated checking of your source code for programmatic and stylistic errors. Most of the time, basic linting is already enabled but it is a good idea to double check.
```

* Line margins - we have code style rules that forbid lines longer than 79 characters, so having a ruler at 79 characters helps coders adhere to that rule.

> * Introspection - this helps removes trailing whitespaces that could be problematic when compiling or building

Reviewer: ```suggestion
* Introspection helps by removing trailing whitespaces that can be problematic when compiling or building.
```

> * Line margins - we have code style rules that forbid lines longer than 79 characters, so having a ruler at 79 characters helps coders adhere to that rule.
> 
> * Introspection - this helps removes trailing whitespaces that could be problematic when compiling or building
> 
> * Docstring - we use NumPy because it supports a combination of reStructured and GoogleDocstrings and supported by Sphinx

Reviewer: ```suggestion
* Line margins - we have code style rules that forbid lines longer than 79 characters, so having a ruler at 79 characters helps coders adhere to that rule.

* Introspection - this helps removes trailing whitespaces that could be problematic when compiling or building

* Docstring - we use NumPy because it supports a combination of reStructured and GoogleDocstrings and is supported by Sphinx
```
I think this bullet point also could have a bit more information.

Reviewer: (the last one on "Docstring" that is - just seeing now GitHub shows the previous ones in the view here too 🙃 )

Spyder
------

* :menuselection:`Preferences --> Completion and Linting --> Linting --> Enable Basic Linting`
* :menuselection:`Preferences --> Completion and Linting --> Introspection --> Advanced -->Add "mne" to list of modules to preload for code completion`
* :menuselection:`Preferences --> Editor --> Show Vertical Lines at == 79`
* :menuselection:`Preferences --> Editor --> Advanced Settings --> Docstring Type == Numpy`

> * :menuselection:`Preferences --> Editor --> Code Introspection/Analysis and “Automatically remove trailing spaces when saving files”`

Reviewer 2: A couple of modification/additions, to match the entries in the latest Spyder 5.4.2.

```
:menuselection:`Preferences --> Editor --> Show Vertical Lines at == 79`
```

This entry has been moved to Completion and linting -> Code style and formatting -> Line length.
In the same section, we could recommend to change the code formatting provider from `autopep8` to `black`.

In Completion and linting -> Docstring style -> Enable docstring style linting and choose the `numpy` convention replaces:
```
:menuselection:`Preferences --> Editor --> Advanced Settings --> Docstring Type == Numpy
```

And finally:

```
:menuselection:`Preferences --> Editor --> Code Introspection/Analysis and “Automatically remove trailing spaces when saving files”
```

is replaced by Editor -> Source code -> Automatically remove trailing spaces when saving files.
In the same tab, we could also recommend to enable: 
- Insert a newline at the end if one does not exist when saving a file
- Trim all newlines after the final one when saving a file

And in the Advaned settings tab -> Convert end-of-line character to the following on save -> LF (Unix).

Reviewer: This actually makes me think that this will be very hard to maintain / keep up to date ... Would it be enough to only keep the general recommendations but not the "how-to" regarding the IDE's?

Reviewer 2: Exactly what I thought when I noticed all the differences..

PyCharm
-------

* :menuselection:`Settings/Preferences (Ctrl+Alt+S) --> Languages and Frameworks --> JavaScript --> Code Quality Tools --> Closure Linter > Enable`
* :menuselection:`View --> Tool Windows --> Python Packages --> Search (Enter Package) --> Select Required Version and press Install`
* :menuselection:`File --> Settings --> Editor --> Code Style --> General: Right Margin (columns) Change to 79`
* :menuselection:`File --> Settings --> Tools --> Python Integrated Tools --> Docstring = NumPy`
* :menuselection:`Settings --> Editors --> General --> scroll to On Save --> change Remove trailing spaces on: to ``Modified Lines```

VsCode
------

* :menuselection:`Command Palette (Ctrl + Shift + P) --> Lint --> Python: Enable/Disable Linting (click on Enable)`

> * :menuselection:`Use ``python.autoComplete.preloadModules": ["numpy", "pandas", "matplotlib", "mne"]```

Reviewer 3: Is this really still a thing with Pylance?

Author: I'm just continuing off from the work that was left unattended for a while. Not sure what versioning was in place when the instructions were last created. Let me know if there needs to be a change.
